base.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+	"github.com/juju/os/v2"
+	"github.com/juju/utils/v2/arch"
+)
+
+// Base represents an OS/Channel.
+// Bases can also be converted to and from a series string.
+type Base struct {
+	Name          string   `json:"name,omitempty"`
+	Channel       Channel  `json:"channel,omitempty"`
+	Architectures []string `json:"architectures,omitempty"`
+}
+
+// Validate returns with no error when the Base is valid.
+func (b Base) Validate() error {
+	if b.Name == "" {
+		return errors.NotValidf("name must be specified")
+	}
+
+	if !validOSForBase.Contains(b.Name) {
+		return errors.NotValidf("os %q", b.Name)
+	}
+	if b.Channel.Empty() {
+		return errors.NotValidf("channel")
+	}
+	for _, v := range b.Architectures {
+		if !arch.IsSupportedArch(v) {
+			return errors.NotValidf("architecture %q", v)
+		}
+	}
+
+	return nil
+}
+
+// String representation of the Base.
+func (b Base) String() string {
+	str := strings.ToLower(b.Name)
+	if !b.Channel.Empty() {
+		str += "/" + b.Channel.String()
+	}
+	if len(b.Architectures) > 0 {
+		str = fmt.Sprintf("%s on %s", str, strings.Join(b.Architectures, ", "))
+	}
+	return str
+}
+
+// ParseBaseFromString parses a base as string in the form
+// "os/track/risk/branch"
+func ParseBaseFromString(s string) (Base, error) {
+	base, err := parseBaseString(s)
+	if err != nil {
+		return Base{}, err
+	}
+	err = base.Validate()
+	if err != nil {
+		return Base{}, errors.Annotatef(err, "invalid base string %q", s)
+	}
+	return base, nil
+}
+
+// ParseBaseWithArchitectures parses a base as string in the form
+// "os/track/risk/branch" and a list of architectures
+func ParseBaseWithArchitectures(s string, archs []string) (Base, error) {
```

Comment:
We may want just one ParseBase, including architectures in the single string would be unwieldy.

Reply:
I think one method with a variadic arg for architectures would work.

Reply:
thanks